base/utfcheck.jl

```diff
@@ -0,0 +1,195 @@
+                    surr = get_continuation(0x0000d, dat[pos += 1], pos)
+                    surr = get_continuation(surr, dat[pos += 1], pos)
+                    !is_surrogate_trail(surr) && throw(UnicodeError(UTF_ERR_NOT_TRAIL, pos-2, surr))
+                    (options & UTF_NO_SURROGATES) != 0 && throw(UnicodeError(UTF_ERR_SURROGATE, pos-2, surr))
+                    flags |= UTF_SURROGATE
+                    num4byte += 1
+                elseif ch > 0x07ff
+                    num3byte += 1
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    flags |= UTF_LONG
+                    num2byte += 1
+                else
+                    throw(UnicodeError(UTF_ERR_LONG, pos-2, ch))
+                end
+            elseif ch < 0xf5
+                # 4-byte UTF-8 sequence (i.e. characters > 0xffff)
+                (pos + 3 > len) && throw(UnicodeError(UTF_ERR_SHORT, pos, ch))
+                ch = get_continuation(ch & 0x07, dat[pos += 1], pos)
+                ch = get_continuation(ch, dat[pos += 1], pos)
+                ch = get_continuation(ch, dat[pos += 1], pos)
+                if ch > 0x10ffff
+                    throw(UnicodeError(UTF_ERR_INVALID, pos-3, ch))
+                elseif ch > 0xffff
+                    num4byte += 1
+                elseif is_surrogate_codeunit(ch)
+                    throw(UnicodeError(UTF_ERR_SURROGATE, pos-3, ch))
+                elseif (options & UTF_ACCEPT_LONG) != 0
+                    # This is an overly long encoded character
+                    flags |= UTF_LONG
+                    if ch > 0x7ff
+                        num3byte += 1
+                    elseif ch > 0x7f
+                        num2byte += 1
+                    end
+                else
+                    throw(UnicodeError(UTF_ERR_LONG, pos-2, ch))
+                end
+            else
+                throw(UnicodeError(UTF_ERR_INVALID, pos, ch))
+            end
+        end
+    end
+    num3byte != 0 && (flags |= UTF_UNICODE3)
+    num4byte != 0 && (flags |= UTF_UNICODE4)
+    return totalchar, flags, num4byte, num3byte, num2byte
+end
+
+" Validates and calculates number of characters in a UTF-16 or UTF-32 encoded vector/string
+
+### Input Arguments:
+* dat    Vector of UInt16, UInt32, or an AbstractString
+
+### Optional Input Arguments:
+* len    length
+* pos    start position
+
+### Keyword Argument:
+* options Determines error handling, default is:
+    * Accept long \0 (Modified UTF-8)
+    * Accept surrogate pairs in UTF-8 (CESU-8)
+    * Don't accept other long encodings
+
+### Returns:
+*   (total characters, flags, 4-byte, 3-byte, 2-byte)
+
+### Throws:
+*   UnicodeError
+"
+function check_string{T <: Union(Vector{UInt16}, Vector{UInt32}, AbstractString)}(
+                      dat::T,
+                      len = endof(dat),
+                      pos = (T <: AbstractString) ? start(dat) : 1
```

Comment:
I don't think such a case exists. When exploring https://github.com/JuliaLang/julia/issues/9297 I even considered allowing `1` as the only accepted integer to index a string, because it was painful to change all code to use `start(s)`.